doc/hooks/versionCheckHook.section.md

```diff
@@ -0,0 +1,35 @@
+# versionCheckHook {#versioncheckhook}
+
+This hook adds a `versionCheckPhase` to the [`preInstallCheckHooks`](#ssec-installCheck-phase) that runs the main program of the derivation with a `--help` or `--version` argument, and checks that the `${version}` string is found in that output. You use it like this:
+
+```nix
+{
+  lib,
+  stdenv,
+  versionCheckHook,
+  # ...
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  # ...
+
+  nativeInstallCheckInputs = [
+    versionCheckHook
+  ];
+  doInstallCheck = true;
+
+  # ...
+})
+```
+
+Note that for [`buildPythonPackage`](#buildpythonpackage-function) and [`buildPythonApplication`](#buildpythonapplication-function), `doInstallCheck` is enabled by default.
+
+It does so in a clean environment (using `env --ignore-environment`), and it checks for the `${version}` string in both the `stdout` and the `stderr` of the command. It will report to you in the build log the output it received and it will fail the build if it failed to find `${version}`.
+
+The variables that this phase control are:
+
+- `dontVersionCheck`: Disable adding this hook to the [`preDistPhases`](#var-stdenv-preDist). Useful if you do want to load the bash functions of the hook, but run them differently.
+- `versionCheckProgram`: The full path to the program that should print the `${version}` string. Defaults roughly to `$out/bin/$pname`.
```

Comment:
"full path" is problematic, or at least very confusing.

I would assume setting `versionCheckProgram = "$out/bin/foobar";` will not expand `$out`, and i'm not sure if `${placeholder "out"}bin/foobar` will work since, the placeholder may be know ahead of the build or after the build is finished. A canonical example of how to set this correctly is much appreciated.

Reply:
> "full path" is problematic, or at least very confusing.
> 
> I would assume setting `versionCheckProgram = "$out/bin/foobar";` will not expand `$out`, 

Correct.

> and i'm not sure if `${placeholder "out"}bin/foobar` will work, since whether the placeholder is known ahead of the build or only after the build is finished may vary.

Really? I thought that once all the inputs to a derivation are known to Nix, it knows the output path of it...

> A canonical example of how to set this correctly is much appreciated.

Using `placeholder` worked for me, so I added an example and mentioned the fact that `$out` is not expanded and `placeholder` will probably always be needed. I also checked this interface like this:

```patch
diff --git a/pkgs/by-name/he/hello/package.nix b/pkgs/by-name/he/hello/package.nix
index 7b8db4c7c3c2..2a1540d8918b 100644
--- a/pkgs/by-name/he/hello/package.nix
+++ b/pkgs/by-name/he/hello/package.nix
@@ -23,6 +23,8 @@ stdenv.mkDerivation (finalAttrs: {
   nativeInstallCheckInputs = [
     versionCheckHook
   ];
+  versionCheckProgram = "${placeholder "out"}/bin/hello";
+  versionCheckProgramArg = "--version";
 
   # Give hello some install checks for testing purpose.
   postInstallCheck = ''
```

And it worked according to the build log.

Reply:
As someone who stumbled into this check being applied, it was confusing to me that `versionCheckProgram` had to be a full path. The placeholder does work.